### DISCUSSION TITLE
Static Reflection Support

### DISCUSSION THREAD

**Discussion by nicolasgustafsson** [carbon-language/carbon-lang#1415] [General]
Static Reflection is near-critical in my field of work(Gamedev), for the ability to annotate member variables and automatic serialization. Current c++ engines mostly rely on hacks to get it working(see [unreal header tool](https://docs.unrealengine.com/4.27/en-US/ProductionPipelines/BuildTools/UnrealHeaderTool/)); scanning through the docs, I do not find a mention of it. 

I realize that this is a very complex language feature and other things have much higher priority, so this is a more general question for the far future: Does the Carbon team have an official stance regarding the inclusion of reflection as a language feature?

**Comment by jwtowner** [comment]
If you search for "metaprogramming," it looks like there are tentative plans to support static and runtime reflection through metaprogramming facilities.

**Reply by nicolasgustafsson** [reply]
Indeed, [this](https://github.com/carbon-language/carbon-lang/pull/89/files#diff-63ec49774932ca9c14564adc104a4aa8b44b172bd0b7761e94aaca2d45e6e5f1=) seems to include reflection.
I also searched the discord and I get the impression that reflection is planned, but in the far future. That's great! As long as there is some plan for inclusion I'm happy. I was a bit scared that it would be a non-goal as I could not find anything regarding it.

**Reply by huiseliming** [reply]
It should be a flexible way to provide traversable Ast programmable macros that run during the preprocessing stage

**Reply by huiseliming** [reply]
This will be very convenient for DSL and reflection, etc.
